src/mysky/index.ts

```diff
@@ -0,0 +1,302 @@
+    }
+
+    const connector = await Connector.init(client, mySkyDomain, customOptions);
+
+    const domain = await client.extractDomain(window.location.hostname);
+    // TODO: Are these permissions correct?
+    const perm = new Permission(domain, skappDomain, PermCategory.Hidden, PermType.Write);
+    const permissions = [perm];
+
+    MySky.instance = new MySky(connector, permissions, domain);
+    return MySky.instance;
+  }
+
+  // ==========
+  // Public API
+  // ==========
+
+  /**
+   * Loads the given DACs.
+   */
+  async loadDacs(...dacs: DacLibrary[]): Promise<void> {
+    const promises: Promise<void>[] = [];
+    for (const dac of dacs) {
+      promises.push(this.loadDac(dac));
+    }
+
+    await Promise.all(promises);
+  }
+
+  async addPermissions(...permissions: Permission[]) {
+    this.pendingPermissions.push(...permissions);
+  }
+
+  async checkLogin(): Promise<boolean> {
+    const failedPermissions: Permission[] = await this.connector.connection
+      .remoteHandle()
+      .call("checkLogin", this.pendingPermissions);
+
+    // Save failed permissions.
+    this.pendingPermissions = failedPermissions;
+
+    if (failedPermissions.length > 0) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Destroys the mysky connection by:
+   *
+   * 1. Destroying the connected DACs,
+   *
+   * 2. Closing the connection,
+   *
+   * 3. Closing the child iframe
+   */
+  async destroy(): Promise<void> {
+    // TODO: For all connected dacs, send a destroy call.
+
+    // TODO: Delete all connected dacs.
+
+    // Close the connection.
+    this.connector.connection.close();
+
+    // Close the child iframe.
+    if (this.connector.childFrame) {
+      this.connector.childFrame.parentNode!.removeChild(this.connector.childFrame);
+    }
+  }
+
+  async logout(): Promise<void> {
+    // TODO
+  }
+
+  async requestLoginAccess(): Promise<boolean> {
+    // Add error listener.
+
+    const { promise: promiseError, controller: controllerError } = this.monitorUiError();
+
+    let uiWindow: Window;
+    let uiConnection: Connection;
+    const promise: Promise<void> = new Promise(async (resolve, reject) => {
+      // Make this promise run in the background and reject on window close or any errors.
+      promiseError.catch((err: string) => {
+        if (err === "closed") {
+          // Resolve without updating the pending permissions.
+          resolve();
+        }
+
+        reject(err);
+      });
+
+      try {
+        // Launch the UI.
+
+        [uiWindow, uiConnection] = await this.launchUi();
+
+        // Send the UI the list of required permissions.
+
+        // TODO: This should be a dual-promise that also calls ping() on an interval and rejects if no response was found in a given amount of time.
+        const failedPermissions: Permission[] = await uiConnection
+          .remoteHandle()
+          .call("requestLoginAccess", this.pendingPermissions);
```

Comment:
I haven't seen a `requestLoginAccess` that returns a list of permissions it was not able to acquire. That's how it works right? Where does that live?

Reply:
`skynet-mysky`, I will push that by the the time you're on tomorrow.